compiler/rustc_passes/src/check_attr.rs

```diff
@@ -989,16 +989,36 @@ impl CheckAttrVisitor<'tcx> {
 
         for hint in &hints {
             let (article, allowed_targets) = match hint.name_or_empty() {
-                _ if !matches!(target, Target::Struct | Target::Enum | Target::Union) => {
-                    ("a", "struct, enum, or union")
+                _ if !matches!(
```

Comment:
this already existed before, but why do we match eagerly here  :thinking: 

this seems unnecessary and slightly worsens diagnostics. I would prefer us to remove that match arm and
put a `delay_span_bug' in the wildcard arm here.

You don't have to do that in this PR if you want though.

Reply:
figured I may as well, so I added this and fixed the few tests that failed, except for one: https://github.com/rust-lang/rust/blob/36f1f04f18b89ba4a999bcfd6584663fd6fc1c5d/src/test/ui/issues/issue-43988.rs#L14-L22 which has two errors not longer printed, I'm assuming because `delay_span_bug` stuff isn't emitted if there are other errors before that? is there a particular way I should resolve that?

Reply:
(friendly poke :D)

Reply:
what happens if we change the last match arm to something like
```rust
_ => struct_span_err!(
                self.tcx.sess,
                hint.span(),
                E0552,
                "unrecognized representation hint"
            )
            .emit();
```
you might have to change `ReprOptions::new` to a `delay_span_bug` instead.

Reply:
I'll give that a go in a couple hours and report back :)

Reply:
changed the final match arm to
```rust
_ => {
    struct_span_err!(
        self.tcx.sess,
        hint.span(),
        E0552,
        "unrecognized representation hint"
    )
    .emit();
    continue;
}
```
which resulted in the following other test failure:
```
error[E0565]: meta item in `repr` must be an identifier
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error[E0565]: meta item in `repr` must be an identifier
  --> /home/repnop/dev/rust/rust/src/test/ui/error-codes/E0565.rs:2:8
   |
LL | #[repr("C")] //~ ERROR E0565
   |        ^^^

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0552, E0565.
For more information about an error, try `rustc --explain E0552`.
```

the output for `issue-43988.rs` now looks like:
```
error: malformed `repr` attribute input
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:24:5
   |
LL |     #[repr]
   |     ^^^^^^^ help: must be of the form: `#[repr(C)]`

error: malformed `repr` attribute input
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:34:14
   |
LL |     let _z = #[repr] 1;
   |              ^^^^^^^ help: must be of the form: `#[repr(C)]`

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:5:5
   |
LL |     #[inline]
   |     ^^^^^^^^^
LL |     let _a = 4;
   |     ----------- not a function or closure

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:10:5
   |
LL |     #[inline(XYZ)]
   |     ^^^^^^^^^^^^^^
LL |     let _b = 4;
   |     ----------- not a function or closure

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:14:12
   |
LL |     #[repr(nothing)]
   |            ^^^^^^^

error[E0552]: unrecognized representation hint
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:18:12
   |
LL |     #[repr(something_not_real)]
   |            ^^^^^^^^^^^^^^^^^^

error[E0518]: attribute should be applied to function or closure
  --> /home/repnop/dev/rust/rust/src/test/ui/issues/issue-43988.rs:30:5
   |
LL |     #[inline(ABC)]
   |     ^^^^^^^^^^^^^^
LL |     foo();
   |     ----- not a function or closure

error: aborting due to 7 previous errors

Some errors have detailed explanations: E0518, E0552.
For more information about an error, try `rustc --explain E0518`.
```
which is a definite win, I think.

as for the change to `ReprOptions::new`, I don't see anything that emits errors in the function, so I'm assuming nothing needs adjusted in it.

Reply:
ah yeah, meant `fn find_repr_attrs` instead of `ReprOptions::new`. If you move the error reporting from there to `check_attrs` this should be a cleaner solution.

So changing the following to a delay_span_bug:
```rust
                if !recognised {
                    // Not a word we recognize
                    struct_span_err!(
                        diagnostic,
                        item.span(),
                        E0552,
                        "unrecognized representation hint"
                    )
                    .emit();
                }
```

and completely moving the `if !item.is_meta_item()`, should make this a lot cleaner i think

Reply:
just to make sure I'm understanding correctly: do you want me to move all of the error reporting from `find_repr_attrs` to `check_attrs` or just those specific instances? :thinking:

Reply:
i think it's easier to just move those specific instances. If you're able to cleanly move all the error reporting to here that's also fine for me

Reply:
okay I _think_ I got it